lightning/src/util/events.rs

```diff
@@ -566,6 +591,33 @@ pub enum Event {
 		/// now + 5*time_forwardable).
 		time_forwardable: Duration,
 	},
+	/// Used to indicate that we've intercepted an HTLC forward. This event will only be generated if
+	/// you've encoded an intercept scid in the receiver's invoice route hints using
+	/// [`ChannelManager::get_intercept_scid`].
+	///
+	/// [`ChannelManager::forward_intercepted_htlc`] or
+	/// [`ChannelManager::fail_intercepted_htlc`] MUST be called in response to this event. See
+	/// their docs for more information.
+	///
+	/// [`ChannelManager::get_intercept_scid`]: crate::ln::channelmanager::ChannelManager::get_intercept_scid
+	/// [`ChannelManager::forward_intercepted_htlc`]: crate::ln::channelmanager::ChannelManager::forward_intercepted_htlc
+	/// [`ChannelManager::fail_intercepted_htlc`]: crate::ln::channelmanager::ChannelManager::fail_intercepted_htlc
+	HTLCIntercepted {
+		/// The fake scid that was programmed as the next hop's scid, generated using
+		/// [`ChannelManager::get_intercept_scid`].
+		///
+		/// [`ChannelManager::get_intercept_scid`]: crate::ln::channelmanager::ChannelManager::get_intercept_scid
+		requested_next_hop_scid: u64,
+		/// The payment hash used for this HTLC.
+		payment_hash: PaymentHash,
+		/// How many msats are to be received on the inbound edge of this HTLC.
+		inbound_amount_msat: u64,
+		/// How many msats the payer intended to route to the next node. Depending on the reason you are
+		/// intercepting this payment, you might take a fee by forwarding less than this amount.
+		expected_outbound_amount_msat: u64,
+		/// A id to help LDK identify which HTLC is being forwarded or failed.
+		intercept_id: InterceptId
```

Comment:
What do we think about introducing some versioning of the `InterceptId` ? In the sense we're just mapping `incoming_shared_secret` to a sha256 image, however we could imagine more scheme in the future, where the application parse onion custom `tlv_record` deterministically generating a given `InterceptId`, of which there is a specific semantic understood only by the application. We could also just give back the onion payload to the application.

I don't know, versioning is protocol design 101 and this `InterceptId` might become some standardized API between Lightning implementations and intercepting applications, therefore versioning reduce odds of collisions and other things.

Reply:
If users are parsing the onion in the future, they could return their own forward ID on onion parse that could be included in intercept events, but it still seems best for LDK to control its internal tracking key so we can ensure it'll be random but still the same across duplicate HTLCs? And their desired ID could just be a separate new field in the event?

Reply:
Yes, I think this approach works well too, just to defer the versioning to the intercepting module, where the forward ID is in the onion parse. Note, I think if we would like intercepting module to work across LN implementations (let's say hold-invoice/escrow or whatever) in the long-term future, this one issue ("versioning-by-the-intercepting-module") that should have common handling, I guess.

Reply:
Okay, I think if we would like to extend the fields of `HTLCIntercept` event effectively we can already new fields and it should be handle well by our serialization framework. Legacy consumers can just ignore them. If we have application which would like to consume the onion payload they can add versioning there (still suppose we add `onion_payload` field in `HTLCIntercept` can be done in the future).